pkg/rbd/rbd_util.go

```diff
@@ -467,141 +1034,252 @@ func protectSnapshot(pOpts *rbdSnapshot, adminID string, credentials map[string]
+		return errors.Wrapf(err, "failed to restore snapshot, command output: %s", string(output))
 	}
 
 	return nil
 }
 
-func deleteSnapshot(pOpts *rbdSnapshot, adminID string, credentials map[string]string) error {
-	var output []byte
+// getSnapshotMetadata fetches on-disk metadata about the snapshot and populates the passed in
+// rbdSnapshot structure
+func getSnapshotMetadata(pSnapOpts *rbdSnapshot, adminID string, credentials map[string]string) error {
+	imageName := pSnapOpts.RbdImageName
+	snapName := pSnapOpts.RbdSnapName
 
-	mon, err := getSnapMon(pOpts, credentials)
+	key, err := getKey(adminID, credentials)
 	if err != nil {
 		return err
 	}
 
-	image := pOpts.VolName
-	snapID := pOpts.SnapID
-
-	key, err := getRBDKey(pOpts.ClusterID, adminID, credentials)
+	snapInfo, err := getSnapInfo(pSnapOpts.Monitors, adminID, key, pSnapOpts.Pool, imageName, snapName)
 	if err != nil {
 		return err
 	}
-	klog.V(4).Infof("rbd: snap rm %s using mon %s, pool %s", image, mon, pOpts.Pool)
-	args := []string{"snap", "rm", "--pool", pOpts.Pool, "--snap", snapID, image, "--id", adminID, "-m", mon, "--key=" + key}
 
-	output, err = execCommand("rbd", args)
+	pSnapOpts.SizeBytes = snapInfo.Size
 
+	tm, err := time.Parse(time.ANSIC, snapInfo.Timestamp)
 	if err != nil {
-		return errors.Wrapf(err, "failed to delete snapshot, command output: %s", string(output))
+		return err
+	}
+
+	pSnapOpts.CreatedAt, err = ptypes.TimestampProto(tm)
+	if err != nil {
+		return err
 	}
 
 	return nil
 }
 
-func restoreSnapshot(pVolOpts *rbdVolume, pSnapOpts *rbdSnapshot, adminID string, credentials map[string]string) error {
-	var output []byte
+// imageInfo strongly typed JSON spec for image info
+type imageInfo struct {
+	ObjectUUID string   `json:"name"`
+	Size       int64    `json:"size"`
+	Format     int64    `json:"format"`
+	Features   []string `json:"features"`
+	CreatedAt  string   `json:"create_timestamp"`
+}
 
-	mon, err := getMon(pVolOpts, credentials)
+// ErrImageNotFound is returned when image name is not found in the cluster on the given pool
+type ErrImageNotFound struct {
+	imageName string
+	err       error
+}
+
+func (e ErrImageNotFound) Error() string {
+	return e.err.Error()
+}
+
+// getImageInfo queries rbd about the given image and returns its metadata, and returns
+// ErrImageNotFound if provided image is not found
+func getImageInfo(monitors, adminID, key, poolName, imageName string) (imageInfo, error) {
+	// rbd --format=json info [image-spec | snap-spec]
+
+	var imgInfo imageInfo
+
+	stdout, _, err := util.ExecCommand(
+		"rbd",
+		"-m", monitors,
+		"--id", adminID,
+		"--key="+key,
+		"-c", util.CephConfigPath,
+		"--format="+"json",
+		"info", poolName+"/"+imageName)
 	if err != nil {
-		return err
+		klog.Errorf("failed getting information for image (%s): (%s)", poolName+"/"+imageName, err)
+		if strings.Contains(string(stdout), "rbd: error opening image "+imageName+
```

Comment:
Why we need this check here ?

Reply:
The check is to detect missing images, and segregate them from other classes of errors. We need this to return success for successive DeleteVolume requests in case the first one times out, but the work is actually done. If we cannot detect image missing errors it would appear to the CO system as internal errors or other errors.

The above is one case of the same, which is needed in the code. The error itself can be used generically to detect that such an image does not exist.